blog/_src/posts/2017-09-27-final-algebra-semantics-is-observational-equivalence.md

```diff
@@ -0,0 +1,317 @@
+This is an equivalent formulation to the definition in Mitch's paper,
+but there it is separated into two properties fullness and
+faithfulness, and doesn't use the initial algebras and \\(m^*\\)
+explicitly.
+
+Now we can verify that the inclusion \\(i : T_0 \to T_1\\) of the number
+theory into the number-multiset theory is an extension in this sense.
+
+Finally we can define our notion of \\(i\\)-algebra, which will be our
+correct notion of algebra.
+An \\(i\\)-algebra is a \\(T_1\\) algebra \\(A\\) such that 
+
+1. The canonical algebra map \\(! : In(T_0) \to m^*A\\) is an isomorphism.
+2. The canonical algebra map \\(! : In(T_1) \to A\\) is surjective i.e.,
+   for each sort \\(s, !_s\\) is surjective.
+
+The first condition says again that we have a conservative extension
+of \\(T_0\\), but the second is more interesting. It says that every
+denotation given by \\(A\\) is represented by some term in \\(T_1\\).  In fact
+what it really ensures is that \\(A\\) determines a *congruence relation*
+on \\(T_1\\) given by \\(t1 \equiv_A t2\\) if \\(A(t1) = A(t2)\\).
+In light of this, the first condition could be called *adequacy*.
+
+Furthermore, the surjectivity condition ensures that any morphism of
+\\(i\\) algebras, i.e., a map as \\(T_1\\)-algebras is also surjective, so a
+morphism \\(A \to B\\) is a witness to the fact that \\(B\\) determines a
+*stronger* congruence relation on \\(T_1\\) than \\(A\\) does: \\(t1 \equiv_B t2
+\implies t1 \equiv_A t2\\).
+Then asking for a final algebra is asking for exactly the
+
+> Strongest adequate congruence relation
+	
+which is exactly the definition of observational equivalence that I
+was taught.
+This formulation is very nice though since it makes obvious that
+*adequacy* is not a predetermined concept, we have to pick \\(T_0\\) and
+\\(i\\) in order to know what adequacy means.
+
+# Conclusion: Tying it back to Final Encodings
+
+So now we've seen that
+
+> Final algebras are equivalent to initial algebras modulo observational equivalence
+
+Of course we haven't precisely gotten back to where we started: we
+were talking about denotational semantics in terms of sets and
+functions, but what we really want are implementations in our favorite
+programming languages.
+Fortunately, we didn't use very many properties of sets in our
+definition, so it's pretty easy to swap out the category of Sets for
+some category built out of the terms of our programming language.
+We can also swap out sets for some much cooler category of denotations
+like domains or metric spaces or time-varying values.
+
+Another question is how to implement this when we have a proper *type
+theory* and not just some boring sorts. In particular, if we have
+function types, then we won't be able to get functions from functions
+in our term model to functions in our denotations due to
+contravariance. Perhaps logical relations are the solution?
+
+[ryanc]: http://ccs.neu.edu/home/ryanc
+[michaelb]: ???
```

Comment:
maybe https://github.com/michaelballantyne

Reply:
Yeah, my github is probably the best right now.